### PR TITLE
loader : Log upsert and remove route errors

### LINKED ISSUE #15282: Log errors on upsert/delete of endpoint routes
When configuring per-endpoint routes, we should log errors:
https://github.com/cilium/cilium/blob/fde0bdac8aa538996ce29cf8275c6f92e0663512/pkg/datapath/loader/loader.go#L348-L356

https://github.com/cilium/cilium/pull/15228 added code to remove the routes when per-endpoint routes are disabled. We should also log errors in that case, unless the error is `ENOENT` when attempting to remove a route that doesn't exist.

### PR DESCRIPTION
**Pull request by h3llix**
Fixes : #15282

Signed-off-by: Gaurav Genani <h3llix.pvt@gmail.com>

### CODE REVIEW

File: pkg/datapath/loader/loader.go [hunk 1]
```diff
@@ -352,17 +352,30 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 
 	if ip := ep.IPv4Address(); ip.IsSet() {
 		if ep.RequireEndpointRoute() {
-			upsertEndpointRoute(ep, *ip.IPNet(32))
+
```
Comment (pchaigno) [inline pkg/datapath/loader/loader.go]: Nit: You can drop the extra newline here. Same for the other three below.

**Comment by pchaigno** [comment]
test-me-please

File: pkg/datapath/loader/loader.go [hunk 2]
```diff
@@ -352,17 +352,30 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
...
+			if err := upsertEndpointRoute(ep, *ip.IPNet(32)); err != nil {
+				log.WithError(err).Error("Failed to upsert route")
```
Comment (pchaigno) [inline pkg/datapath/loader/loader.go]: I think we should use the `scopedLog` here to report the endpoint.

**Review by pchaigno — COMMENTED** [review]
Thanks for your contribution! :rocket:

File: pkg/datapath/loader/loader.go (hunk 2 above)
Comment (h3llix) [reply pkg/datapath/loader/loader.go]: What should be the fields under logrus ?

Comment (pchaigno) [reply pkg/datapath/loader/loader.go]: No new field needed. See previous uses of the `scopedLog` variable above in the same file.

**Comment by pchaigno** [comment]
test-gke

File: pkg/datapath/loader/loader.go [hunk 3]
```diff
@@ -351,18 +351,34 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 	}
 
 	if ip := ep.IPv4Address(); ip.IsSet() {
+		scopedLog := ep.Logger(Subsystem).WithFields(logrus.Fields{
+			logfields.Path: objPath,
```
Comment (pchaigno) [inline pkg/datapath/loader/loader.go]: Ah, sorry, I think I just understood your previous question :facepalm: 

We can drop the `Path` logfield I think. It isn't relevant to endpoint routes.

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Comment by h3llix** [comment]
@pchaigno Thank You for the help .

**Review by christarazi — CHANGES_REQUESTED** [review]
Changes LGTM, please squash the commits together

**Review by christarazi — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

File: pkg/datapath/loader/loader.go [hunk 4]
```diff
@@ -351,18 +351,32 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 	}
 
 	if ip := ep.IPv4Address(); ip.IsSet() {
+		scopedLog := ep.Logger(Subsystem).WithFields(logrus.Fields{
+			logfields.Veth: ep.InterfaceName(),
+		})
 		if ep.RequireEndpointRoute() {
-			upsertEndpointRoute(ep, *ip.IPNet(32))
+			if err := upsertEndpointRoute(ep, *ip.IPNet(32)); err != nil {
+				scopedLog.WithError(err).Warn("Failed to upsert route")
```
Comment (aanm) [inline pkg/datapath/loader/loader.go:359]: @pchaigno these are warnings, which steps can users perform if they see these?

Comment (pchaigno) [reply pkg/datapath/loader/loader.go:359]: Same as for other similar warnings (e.g., warning when we fail to load the BPF program for an endpoint): they should investigate and/or report the warning, but it's probably not blocking/fatal for the Cilium agent.

**Comment by pchaigno** [comment]
test-1.20-4.9

**Comment by pchaigno** [comment]
test-gke

**Comment by nathanjsweet** [comment]
#14608 #15128 #14948 #15425 #15499 all known or seen-before flakes.

**Comment by pchaigno** [comment]
> #14608 #15128 #14948 #15425 #15499 all known or seen-before flakes.

@nathanjsweet One of those failures was legit and the changes ended up breaking master. I think in general we should avoid using closed flake issues as a validation that something is a flake.

**Pull request merged**